File: src/apiClient.ts
import * as fs from 'fs';
import { FetchError, Response } from 'node-fetch';
import * as os from 'os';
import * as path from 'path';
import { pipeline } from 'stream';
import * as util from 'util';
import { CacheManager } from './cacheManager';
import { cacheFeature } from './cli';
import { IAppImage, IFlathub, IApiClient, ICacheManager, ISnap } from './dataStructure';
import { errorMessage, successfullMessage } from './helpers';

const fetch = require('node-fetch').default;


const streamPipeline = util.promisify(pipeline);

/**
 * A wrapper for node-fetch library.
 * @class ApiClient
 */
export class ApiClient implements IApiClient {
  private flathubApi: string;
  private snapApi: string;
  private appimageApi: string;

  private flathubData: IFlathub[];
  private snapData: ISnap;
  private appimageData: IAppImage;
  private cacheManager: ICacheManager;

  constructor() {
    this.cacheManager = new CacheManager();
    this.flathubApi = 'https://flathub.org/api/v1/apps/';
    this.snapApi = 'https://raw.githubusercontent.com/MuhammedKpln/chob/master/snapcraft.json';
    this.appimageApi = 'https://appimage.github.io/feed.json';


    if (cacheFeature && !this.cacheManager.hasCachedSources) {
      errorMessage('⚡ Could not find any cached sources, your search will be cached after this results.');
    }

  }

  public async get(url: string, options: Object = {}): Promise<Response> {
    return new Promise(async (resolve, reject) => {
      const response: Response = await fetch(url, options);
      if (response.status !== 200) {
        return reject(response);
      }

      return resolve(response);
    });
  }

  public async download(url: string, fileName: string): Promise<boolean> {
    const homeFolder = path.resolve(os.homedir(), 'chob');
    const filePath = path.resolve(
      homeFolder,
      `${fileName.toLowerCase()}`,
    );
    if (!fs.existsSync(homeFolder)) {
      fs.mkdirSync(path.resolve(os.homedir(), 'chob'));
    }

    if (fs.existsSync(filePath)) {
      errorMessage(`You already have downloaded this file in ${homeFolder}`);

      return false;
    }
    fs.writeFileSync(filePath, '');

    const file = fs.createWriteStream(filePath);

    const response = await this.get(url);

    return streamPipeline(response.body, file).then(() => {
      successfullMessage(
        `Your download is finished, it is stored at ${filePath}`,
      );

      return true;
    }).catch(() => {
      errorMessage('Could not download the app image.');

      return false;
    });

  }

  public grabDataFromFlathub(): Promise<IFlathub[]> {
    return new Promise(async (resolve, reject) => {

      if (cacheFeature && this.cacheManager.hasCachedSources) {
        const { flathubData } = this.cacheManager.getSourcesFromCache();

        return resolve(flathubData);
      }

      this.get(this.flathubApi).then((resp: Response) => {
        return resp.json();
      }).then((json: IFlathub[]) => {
        this.flathubData = json;

        return resolve(this.flathubData);
      }).catch((err: FetchError) => {
        return reject(err);
      });

    });
  }

  public grabDataFromSnap(): Promise<ISnap> {
    return new Promise<ISnap>((resolve, reject) => {

      if (cacheFeature && this.cacheManager.hasCachedSources) {
        const { snapData } = this.cacheManager.getSourcesFromCache();

        return resolve(snapData);
      }

      this.get(this.snapApi).then((resp: Response) => {
        return resp.json();
      }).then((json: ISnap) => {
        this.snapData = json;

        return resolve(this.snapData);
      }).catch((err: FetchError) => {
        return reject(err);
      });

    });
  }

  public grabDataAppImage(): Promise<IAppImage> {
    return new Promise((resolve, reject) => {
      if (cacheFeature && this.cacheManager.hasCachedSources) {
        const { appimageData } = this.cacheManager.getSourcesFromCache();

        return resolve(appimageData);
      }

      this.get(this.appimageApi).then((resp: Response) => {
        return resp.json();
      }).then((json: IAppImage) => {
        this.appimageData = json;

        return resolve(this.appimageData);
      }).catch((err: FetchError) => {
        return reject(err);
      });

    });
  }
}

File: src/api_client.py
from genericpath import exists
from cache_manager import CacheManager
from helpers import error_message
from type.appimage import AppImage
from typing import List
from type.snap import Snap
from type.flathub import FlatHub
import requests
from requests.models import Response
from os import path
from pathlib import Path


class ApiClient:
    flathub_api: str
    snap_api: str
    appimage_api: str
    flathub_data: List[FlatHub]
    snap_data: List[Snap]
    appimage_data: List[AppImage]
    cache_manager: CacheManager = CacheManager()

    def __init__(self):
        self.flathub_api = 'https://flathub.org/api/v1/apps/'
        self.snap_api = 'https://raw.githubusercontent.com/MuhammedKpln/chob/master/snapcraft.json'
        self.appimage_api = 'https://appimage.github.io/feed.json'

    def get(self, url: str, *args, **kwargs) -> Response:
        request = requests.get(url, *args, **kwargs)

        if request.status_code != 200:
            raise ConnectionAbortedError(
                'The URL did not return 200 status code.')

        return request

    def get_flathub_data(self) -> None:
        if self.cache_manager.is_cache_enabled and self.cache_manager.has_cached_sources:

            self.flathub_data = self.cache_manager.fetch_sources_from_cache()[
                'flathub_data']
            return None

        api = self.get(self.flathub_api)

        self.flathub_data = api.json()

    def get_snap_data(self) -> None:
        if self.cache_manager.is_cache_enabled and self.cache_manager.has_cached_sources:
            self.snap_data = self.cache_manager.fetch_sources_from_cache()[
                'snap_data']
            return None

        api = self.get(self.snap_api)
        self.snap_data = api.json()['_embedded']['clickindex:package']

    def get_appimage_data(self) -> None:
        if self.cache_manager.is_cache_enabled and self.cache_manager.has_cached_sources:
            self.appimage_data = self.cache_manager.fetch_sources_from_cache()[
                'appimage_data']
            return None

        api = self.get(self.appimage_api)

        self.appimage_data = api.json()['items']

    def download(self, url: str, file_name: str) -> bool:
        home_folder = path.join(Path.home(), 'chob')
        file_path = path.join(home_folder, file_name.lower())

        if exists(file_path):
            error_message(
                f'You already have downloaded this file in {home_folder}')
            return False

        try:
            with open(file_path, 'wb') as f:
                request: Response = requests.get(url, stream=True)
                content = request.iter_content(chunk_size=1024)

                for chunk in content:
                    f.write(chunk)

                f.close()

                return True
        except Exception as err:
            error_message(str(err))
            return False
